### ISSUE TITLE
Directory and file permissions not consistent

### ISSUE THREAD

**Issue by judgej** [salesagility/SuiteCRM#4142]
The default file permissions include read permission for "other" for files, but no permissions at all for "other" for directories. If there is a need for files to be read by "other", then the directories should also need to be accessible.

#### Issue
I found this on CentOS using FastCGI over Apache and trying to publish a custom module.

#### Expected Behavior
Publishing a module should export a zip file.

#### Actual Behavior
I got a "no permissions" error instead of the zip file.

#### Possible Fix
The default file permissions are 493 decimal, or 755 octal. The zip file was *actually* created with 644 octal permissions, so I guess the module is overriding the default file permissions there.

The zip directory default permissions is 1528 decimal, or 2770 octal. The directory was being created with those permissions (drwxrws--- with a group sticky bit). The FastCGI user runs as the same user as the file and directory ownership. It has to, because the FastCGI user created the directories and that is what permissions it gave them. However, possibly apache is using a different user to navigate the directory structure before downloading the zip file, hence the permissions error.

So adding read and execute to the directories fixed the problem. This changes the default directory permissions from 2770 to 2775 octal. Not interested in what that is in decimal ;-) Whether the sticky bit is needed or not, I don't know, but it seems to have no effect either way.

Whether that fix should always be applied or needs to go into the installation docs as a change needed in certain use-cases, I don't know either. It may just be me, or may be a more widespread issue.

The same permissions are used to create other directories too, but unless they need direct web access, then they aren't causing an issue. For example `cache/jsLanguage/en_us.js` is not accessible through the browser due to `drwxrws---` permissions on `jsLanguage`.

#### Steps to Reproduce

1. CentOS, Apache, FastCGI - create a module in Module Builder and try to publish it.
2.
3.
4.

#### Context
If I set default directory permissions to 2775 in the config file before installing, then everything runs fine. I'm unclear how - if at all - this is affecting other people. My old-old SugarCRM install uses 2777 octal directory permissions, but that's running as mod_php so needs those extra permissions.

I've not tried this with nginx yet. Not sure what Unix users run there to navigate the directory structure and execute scripts.

The tl;dr is that the default directory permissions set no permissions for "other" and that makes all directories created by the web user (and consequently owned by the web user) inaccessible through the web. Adding read+execute to "other" fixes this.

#### Your Environment

* SuiteCRM Version used: 7.8.x
* Browser name and version (e.g. Chrome Version 51.0.2704.63 (64-bit)):
* Environment name and version (e.g. MySQL, PHP 7): PHP7.1
* Operating System and version (e.g Ubuntu 16.04):

**Comment by pgorod** [comment]
Hi @judgej this is a complicated issue because there are inconsistencies in the way the default_permissions in config.php are applied - often they simply are not applied at all. See #2810 for details.

But I'm curious why you would need the "other" access. What user is your Apache running under? And which user/group owns the files? Thanks

**Comment by judgej** [comment]
The files and directories are owned by the web user, which is the same as the FTP user, since it is running FastCGI under Plesk. I can only imagine that perhaps apache navigates the directory structure first, interpreting the `.htaccess` files, then switches to the script/file/directory owner once it gets there. If that's not how it works, then I'm a little confused by this too. PHP scripts can access that directory fine, but apache cannot deliver files from that directory.

Thinking about it, FastCGI is just the way PHP is run, and defines the user the PHP scripts run as. So if downloading a zip file directly from a directory, PHP is *not involved at all*, so the user accessing that file will be apache, which is neither the owner nor in the same group as the directory, hence "other" needs read and execute access to those directories. If the zip file was streamed through a PHP script, then there would be no problems downloading it.

**Comment by judgej** [comment]
I wonder if Plesk has other access modes, with the whole request being handled by the same user? I'll investigate.

**Comment by pgorod** [comment]
That all sounds quite strange. But it should be testable, so maybe we can get to the bottom of this.

There is a problem I see occurring with SuiteCRM occasionally, which is related to the SetGID bit. I've seen cases where the directory has generous permissions but the stuff created under it gets less generous permissions because of the changes in ownership. Setting the SetGID bit solves this.

Contrary to all current SugarCRM/SuiteCRM permissions wisdom, I am thinking we should completely remove permissions from "other", always give them zero. If you get the ownerships right, both user and group, you don't need those concessions. On the other hand, if you get them wrong, the "other" permissions normally won't save you, because they are still insufficient (read-only), except for people who really drop all security and go for 777...

**Comment by gunnicom** [comment]
I like the rights for "group", because of editing is usually not done as "www-data" user.

**Comment by pgorod** [comment]
While this issue is essentially correct, and an interesting thing to improve eventually, there are other issues that involve the same problem and that have more information. So this is somewhat of a duplicate, although not quite.

https://github.com/salesagility/SuiteCRM/issues/2810
https://github.com/salesagility/SuiteDocs/issues/275

I hope and believe that when we do advance this permissions issue, the `other` permissions will be gone for good. So this will be taken care of.

I suggest that we close this one, while thanking @judgej and all those involved for the discussion.
